base/system/services/database.c

```diff
@@ -621,9 +621,50 @@ ScmCreateOrReferenceServiceImage(PSERVICE pService)
 VOID
 ScmRemoveServiceImage(PSERVICE_IMAGE pServiceImage)
 {
+    DWORD dwError;
+
     DPRINT1("ScmRemoveServiceImage() called\n");
 
-    /* FIXME: Terminate the process */
+    /*
+     * No services are running in this image anymore.
+     * Tell the service dispatcher to exit now, and wait
+     * for the process to cleanly terminate.
+     */
+// FIXME: Re-enable here and remove the L"" SERVICE_CONTROL_STOP call in rpcserver.c
+#if 0
+    dwError = ScmControlService(pServiceImage->hControlPipe,
+                                L"",
+                                NULL,
+                                SERVICE_CONTROL_STOP); // Windows: use SERVICE_STOP
+    if (dwError == ERROR_SUCCESS)
+#endif
+    {
+        dwError = WaitForSingleObject(pServiceImage->hProcess, 20000);
+        if (dwError != WAIT_OBJECT_0)
+            DPRINT1("WaitForSingleObject failed, going to kill the process.\n");
+    }
+#if 0
+    else
+    {
+        DPRINT1("ScmControlService failed, going to kill the process.\n");
+    }
+#endif
+
+    if (dwError != ERROR_SUCCESS || dwError != WAIT_OBJECT_0)
+    {
+        /* If the control or the wait failed, kill the host process
+         * (asynchronously), unless it is ourselves (case of services.exe
+         * hosting services) */
+        if (pServiceImage->hProcess != GetCurrentProcess())
+        {
+            TerminateProcess(pServiceImage->hProcess, 0);
+
+            /* Be sure the process really terminates */
+            dwError = WaitForSingleObject(pServiceImage->hProcess, 20000);
+        }
+        if (dwError != WAIT_OBJECT_0)
+            DPRINT1("WaitForSingleObject failed, the process cannot be killed.\n");
```

Comment:
Do you mean
```suggestion
            if (dwError != WAIT_OBJECT_0)
                DPRINT1("WaitForSingleObject failed, the process cannot be killed.\n");
        }
```
?

I don't know if you want to add an `else "Cannot kill current process"`.

Then, maybe add a comment to tell what happens next if the process could not be stopped nor killed:
is it fine to do the following existing cleanup anyway? is the process going to crash or nicely "disappear"? ...

Reply:
The process might crash, but at this point I would say, "we don't care", since we tried to kill it already.